### PR TITLE
Fix GenericApplicationContextTests on Microsoft Windows

### PR DESCRIPTION
**Pull request by jasonjiang9527**
origin unit test will cause bug when get a resource with unknown protocol 

Related to:

- #28703 
- commit 9868c28c737cfadc50f029887e029155b2b2085d

### CODE REVIEW

**Comment by sbrannen** [comment]
> origin unit test will cause bug when get a resource with unknown protocol

The unit test in question passes unmodified.

Can you please explain why you think this is a bug?

**Comment by 819635822** [comment]
> 

here is the origin code : 
```
	private void assertGetResourceSemantics(ResourceLoader resourceLoader, Class<? extends Resource> defaultResouceType) {
		if (resourceLoader != null) {
			context.setResourceLoader(resourceLoader);
		}

		String pingLocation = "ping:foo";
		String fileLocation = "file:foo";

		Resource resource = context.getResource(pingLocation);
		assertThat(resource).isInstanceOf(defaultResouceType);
		resource = context.getResource(fileLocation);
		assertThat(resource).isInstanceOf(FileUrlResource.class);

		context.addProtocolResolver(new PingPongProtocolResolver());

		resource = context.getResource(pingLocation);
		assertThat(resource).asInstanceOf(type(ByteArrayResource.class))
			.extracting(bar -> new String(bar.getByteArray(), UTF_8))
			.isEqualTo("pong:foo");
		resource = context.getResource(fileLocation);
		assertThat(resource).isInstanceOf(FileUrlResource.class);
	}
```

this line  `Resource resource = context.getResource(pingLocation);` will throw a InvalidPathException , just because context don't know how to resolve the `ping` protocol yet

**Comment by 819635822** [comment]
and the stacktrace is 
```
Illegal char <:> at index 4: ping:foo
java.nio.file.InvalidPathException: Illegal char <:> at index 4: ping:foo
	at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
	at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
	at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:232)
	at java.base/java.io.File.toPath(File.java:2387)
	at org.springframework.core.io.FileSystemResource.<init>(FileSystemResource.java:83)
	at org.springframework.core.io.FileSystemResourceLoader$FileSystemContextResource.<init>(FileSystemResourceLoader.java:65)
	at org.springframework.core.io.FileSystemResourceLoader.getResourceByPath(FileSystemResourceLoader.java:54)
	at org.springframework.core.io.DefaultResourceLoader.getResource(DefaultResourceLoader.java:170)
	at org.springframework.context.support.GenericApplicationContext.getResource(GenericApplicationContext.java:245)
	at org.springframework.context.support.GenericApplicationContextTests.assertGetResourceSemantics(GenericApplicationContextTests.java:256)
	at org.springframework.context.support.GenericApplicationContextTests.getResourceWithCustomResourceLoader(GenericApplicationContextTests.java:245)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy1/jdk.proxy1.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)


```

**Comment by sbrannen** [comment]
Ahhhh... then it appears to be an issue when running the test on Microsoft Windows, since that error does not occur on Mac OS or Linux.

Thanks for raising the issue!

File: spring-context/src/test/java/org/springframework/context/support/GenericApplicationContextTests.java [hunk 1]
```diff
@@ -249,11 +249,12 @@ private void assertGetResourceSemantics(ResourceLoader resourceLoader, Class<? e
 		if (resourceLoader != null) {
 			context.setResourceLoader(resourceLoader);
 		}
-
```
Comment (sbrannen) [inline spring-context/src/test/java/org/springframework/context/support/GenericApplicationContextTests.java:252]: Please revert the unnecessary formatting change.

File: spring-context/src/test/java/org/springframework/context/support/GenericApplicationContextTests.java [hunk 2]
```diff
@@ -262,8 +263,11 @@ private void assertGetResourceSemantics(ResourceLoader resourceLoader, Class<? e
 
 		resource = context.getResource(pingLocation);
 		assertThat(resource).asInstanceOf(type(ByteArrayResource.class))
-			.extracting(bar -> new String(bar.getByteArray(), UTF_8))
-			.isEqualTo("pong:foo");
+				.extracting(bar -> new String(bar.getByteArray(), UTF_8))
+				.isEqualTo("pong:foo");
```
Comment (sbrannen) [inline spring-context/src/test/java/org/springframework/context/support/GenericApplicationContextTests.java:267]: Please revert the unnecessary formatting change.

File: spring-context/src/test/java/org/springframework/context/support/GenericApplicationContextTests.java [hunk 3]
```diff
@@ -249,11 +249,12 @@ private void assertGetResourceSemantics(ResourceLoader resourceLoader, Class<? e
 		if (resourceLoader != null) {
 			context.setResourceLoader(resourceLoader);
 		}
-
+		String commonLocation = "foo";
```
Comment (sbrannen) [inline spring-context/src/test/java/org/springframework/context/support/GenericApplicationContextTests.java:252]:
```suggestion
		String relativePathLocation = "foo";
```

Please rename the variable.

**Review by sbrannen — CHANGES_REQUESTED** [review]
Please make the requested changes, and I'll perform a follow-up review before merging.

**Comment by sbrannen** [comment]
This has been merged into `5.3.x` and `main` in  59180e76a7b35f060c056c2f44edc59d6e63a6ca.

Thanks!

----

Please note that I revised the approach in e608b367135fe232d359bd878085e609d2465a17 so that we are actually testing what happens with a `ping:` protocol before the `PingPongProtocolResolver` is registered, which is one of the main goals of those tests.

So, @jasonjiang9527, if you could confirm that the revised tests also pass on your Windows machine I would be very grateful.

**Comment by jasonjiang9527** [comment]
@sbrannen 
the test `getResourceWithDefaultResourceLoader` gets a problem after you commit e608b367135fe232d359bd878085e609d2465a17 .
code:
```
        //not pass  
	@Test
	void getResourceWithDefaultResourceLoader() {
		assertGetResourceSemantics(null, ClassPathResource.class);
	}
        //pass
	@Test
	void getResourceWithCustomResourceLoader() {
		assertGetResourceSemantics(new FileSystemResourceLoader(), FileSystemResource.class);
	}

	private void assertGetResourceSemantics(ResourceLoader resourceLoader, Class<? extends Resource> defaultResourceType) {
		if (resourceLoader != null) {
			context.setResourceLoader(resourceLoader);
		}
                //...
		if (OS.WINDOWS.isCurrentOs()) {
			// On Windows we expect an error similar to the following.
			// java.nio.file.InvalidPathException: Illegal char <:> at index 4: ping:foo
			assertThatExceptionOfType(InvalidPathException.class)
                                //exception here
				.isThrownBy(() -> context.getResource(pingLocation))
				.withMessageContaining(pingLocation);
		}
		else {
			resource = context.getResource(pingLocation);
			assertThat(resource).isInstanceOf(defaultResourceType);
		}

		//...
	}

```
```
	public ClassPathResource(String path, @Nullable ClassLoader classLoader) {
		Assert.notNull(path, "Path must not be null");
		String pathToUse = StringUtils.cleanPath(path);
		if (pathToUse.startsWith("/")) {
			pathToUse = pathToUse.substring(1);
		}
		this.path = pathToUse;
		this.classLoader = (classLoader != null ? classLoader : ClassUtils.getDefaultClassLoader());
	}
```

here is my explation :
if you don't set `ResourceLoader` for `context` , `context` will use `DefaultResourceLoader` resolve the filePath as `ClassPathContextResource` , but `new ClassPathContextResource(path,classLoader)` does not check this path is or not `a real file`  .

with a `FileSystemResourceLoader` , it check the filePath immediately
```
	public FileSystemResource(String path) {
		Assert.notNull(path, "Path must not be null");
		this.path = StringUtils.cleanPath(path);
                //immediately
		this.file = new File(path);
		this.filePath = this.file.toPath();
	}
```
maybe it need to consider whether keep same construct logic between different `Resource` type

**Comment by sbrannen** [comment]
Hi @jasonjiang9527,

Thanks a lot for trying it out and providing detailed feedback!

I'll see if I can fix it once and for all, and I apologize for the inconvenience.

**Comment by jasonjiang9527** [comment]
@sbrannen 
Thanks a lot for your grammar corrections too.

**Comment by jasonjiang9527** [comment]
@sbrannen 
Thanks a lot for your grammar corrections too.

**Comment by sbrannen** [comment]
> Maybe it needs to consider whether to keep the same construct logic between different `Resource` types.

I updated the logic in d1d6eb095e5af75ac80bd5050bf2c50f9ce97077.

Please let me know if you still run into any issues on Windows.

**Comment by jasonjiang9527** [comment]
@sbrannen All tests of the main branch have passed

**Comment by sbrannen** [comment]
> @sbrannen All tests of the main branch have passed

Great!

Thanks for letting me know.

**Pull request closed without merging**